File: vripper-server/src/main/java/tn/mnlr/vripper/services/PathService.java
package tn.mnlr.vripper.services;

import lombok.Getter;
import lombok.NonNull;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import tn.mnlr.vripper.exception.RenameException;
import tn.mnlr.vripper.jpa.domain.Post;
import tn.mnlr.vripper.services.domain.Settings;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;
import java.util.concurrent.locks.ReentrantLock;

@Service
@Slf4j
public class PathService {

  private final DataService dataService;

  @Getter private final ReentrantLock directoryAccess = new ReentrantLock();

  public PathService(DataService dataService) {
    this.dataService = dataService;
  }

  public final File calcDownloadDirectory(Post post, Settings settings) {
    return new File(settings.getDownloadPath(), post.getDownloadDirectory());
  }

  private File getRootFolder(
      @NonNull String forum, @NonNull String threadTitle, Settings settings) {
    File sourceFolder =
        settings.getSubLocation()
            ? new File(settings.getDownloadPath(), sanitize(forum))
            : new File(settings.getDownloadPath());
    return settings.getThreadSubLocation() ? new File(sourceFolder, threadTitle) : sourceFolder;
  }

  public final void createDefaultPostFolder(Post post, Settings settings) {
    File downloadDirectory = getRootFolder(post.getForum(), post.getThreadTitle(), settings);
    downloadDirectory =
        new File(
            downloadDirectory,
            settings.getAppendPostId()
                ? sanitize(post.getTitle()) + "_" + post.getPostId()
                : sanitize(post.getTitle()));
    downloadDirectory = makeDir(downloadDirectory);
    post.setDownloadDirectory(
        downloadDirectory.getAbsolutePath().replace(settings.getDownloadPath(), ""));
    dataService.updatePostDownloadDirectory(post.getDownloadDirectory(), post.getId());
  }

  public final void rename(@NonNull String postId, @NonNull String altName, Settings settings)
      throws RenameException {
    Post post = dataService.findPostByPostId(postId).orElseThrow();
    if (altName.equals(post.getTitle())) {
      return;
    }

    // Download have not started yet
    if (post.getDownloadDirectory() == null) {
      post.setTitle(altName);
      dataService.updatePostTitle(post.getTitle(), post.getId());
      return;
    }

    File newDownloadDirectory = getRootFolder(post.getForum(), post.getThreadTitle(), settings);
    newDownloadDirectory = new File(newDownloadDirectory, sanitize(altName));
    File currentDownloadDirectory = calcDownloadDirectory(post, settings);
    try {
      directoryAccess.lock();
      Files.move(
          currentDownloadDirectory.toPath(),
          newDownloadDirectory.toPath(),
          StandardCopyOption.ATOMIC_MOVE);
      post.setDownloadDirectory(
          newDownloadDirectory.getAbsolutePath().replace(settings.getDownloadPath(), ""));
      dataService.updatePostDownloadDirectory(post.getDownloadDirectory(), post.getId());

      post.setTitle(altName);
      dataService.updatePostTitle(post.getTitle(), post.getId());
    } catch (IOException e) {
      throw new RenameException(
          String.format(
              "Failed to move files from %s to %s", currentDownloadDirectory, newDownloadDirectory),
          e);
    } finally {
      directoryAccess.unlock();
    }
  }

  private File makeDir(@NonNull final File sourceFolder) {
    int counter = 1;
    File folder = sourceFolder;

    while (folder.exists()) {
      folder = new File(sourceFolder.toString() + '.' + counter++);
    }

    if (!folder.mkdirs()) {
      throw new RuntimeException(String.format("Failed to create the folder %s", sourceFolder));
    }

    return folder;
  }

  private String sanitize(final String folderName) {
    String sanitizedFolderName =
        folderName.replaceAll("\\.|\\\\|/|\\||:|\\?|\\*|\"|<|>|\\p{Cntrl}", "_");
    log.debug(String.format("%s sanitized to %s", folderName, sanitizedFolderName));
    return sanitizedFolderName;
  }

  /**
   * Will sanitize the image name and remove extension
   *
   * @param imageName path string
   * @return Sanitized local path string
   */
  public final String formatImageFileName(String imageName) {
    return sanitize(imageName);
  }
}

File: vripper-core/src/main/kotlin/me/mnlr/vripper/services/PathService.kt
package me.mnlr.vripper.services

import org.springframework.stereotype.Service
import me.mnlr.vripper.delegate.LoggerDelegate
import me.mnlr.vripper.model.Settings
import java.io.File
import java.nio.file.Path

@Service
class  PathService {
    private val log by LoggerDelegate()

    fun calculateDownloadPath(forum: String, threadTitle: String, postTitle: String, postId: String, settings: Settings): Path {
        var downloadDirectory =
            if (settings.downloadSettings.forumSubfolder) Path.of(settings.downloadSettings.downloadPath, sanitize(forum)) else Path.of(
                settings.downloadSettings.downloadPath
            )
        downloadDirectory = if (settings.downloadSettings.threadSubLocation) downloadDirectory.resolve(threadTitle) else downloadDirectory
        downloadDirectory = downloadDirectory.resolve(if (settings.downloadSettings.appendPostId) "${sanitize(postTitle)}_${postId}" else sanitize(
            postTitle
        ))
        return downloadDirectory
    }

    private fun makeDir(sourceFolder: File): File {
        var counter = 1
        var folder = sourceFolder
        while (folder.exists()) {
            folder = File(sourceFolder.toString() + '.' + counter++)
        }
        if (!folder.mkdirs()) {
            throw RuntimeException(String.format("Failed to create the folder %s", sourceFolder))
        }
        return folder
    }

    /**
     * Will sanitize the image name and remove extension
     *
     * @param path
     * @return Sanitized local path string
     */
    fun sanitize(path: String): String {
        val sanitizedPath =
            path.replace("\\.|\\\\|/|\\||:|\\?|\\*|\"|<|>|\\p{Cntrl}".toRegex(), "_")
        log.debug(String.format("%s sanitized to %s", path, sanitizedPath))
        return sanitizedPath
    }
}